File: Sources/XMTPiOS/Codecs/GroupUpdatedCodec.swift
//
//  GroupMembershipChanged.swift
//  
//
//  Created by Pat Nakajima on 2/1/24.
//

import Foundation
import LibXMTP

public typealias GroupUpdated = Xmtp_Mls_MessageContents_GroupUpdated

public let ContentTypeGroupUpdated = ContentTypeID(authorityID: "xmtp.org", typeID: "group_updated", versionMajor: 1, versionMinor: 0)

public struct GroupUpdatedCodec: ContentCodec {
	public typealias T = GroupUpdated

	public init() {	}

	public var contentType = ContentTypeGroupUpdated

	public func encode(content: GroupUpdated) throws -> EncodedContent {
		var encodedContent = EncodedContent()

		encodedContent.type = ContentTypeGroupUpdated
		encodedContent.content = try content.serializedData()

		return encodedContent
	}

	public func decode(content: EncodedContent) throws -> GroupUpdated {
		return try GroupUpdated(serializedData: content.content)
	}

	public func fallback(content: GroupUpdated) throws -> String? {
		return nil
	}

	public func shouldPush(content: GroupUpdated) throws -> Bool {
		false
	}
}

File: sdks/android/library/src/main/java/org/xmtp/android/library/codecs/GroupUpdatedCodec.kt
package org.xmtp.android.library.codecs

import org.xmtp.proto.mls.message.contents.TranscriptMessages.GroupUpdated

typealias GroupUpdated = GroupUpdated

val ContentTypeGroupUpdated =
    ContentTypeIdBuilder.builderFromAuthorityId(
        "xmtp.org",
        "group_updated",
        versionMajor = 1,
        versionMinor = 0,
    )

data class GroupUpdatedCodec(
    override var contentType: ContentTypeId = ContentTypeGroupUpdated,
) : ContentCodec<GroupUpdated> {
    override fun encode(content: GroupUpdated): EncodedContent =
        EncodedContent
            .newBuilder()
            .also {
                it.type = ContentTypeGroupUpdated
                it.content = content.toByteString()
            }.build()

    override fun decode(content: EncodedContent): GroupUpdated = GroupUpdated.parseFrom(content.content)

    override fun fallback(content: GroupUpdated): String? = null

    override fun shouldPush(content: GroupUpdated): Boolean = false
}
